File: mirea/cources/2026-2027/aii-july/04/Part-02-Prefix-and-suffix-approach/examples/example-04.py
# Пример: минимум на каждом суффиксе.
# Ввод: n и массив чисел.
# Вывод: массив минимумов на суффиксах.

n = int(input())
a = list(map(int, input().split()))

suf_min = [0] * n
suf_min[-1] = a[-1]
for i in range(n - 2, -1, -1):
    suf_min[i] = min(a[i], suf_min[i + 1])

print(*suf_min)

File: mirea/cources/2026-2027/aii-july/04/Part-02-Prefix-and-suffix-approach/examples/example-04.cpp
// Пример: минимум на каждом суффиксе.
// Ввод: n и массив чисел.
// Вывод: массив минимумов на суффиксах.

#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;

    vector<int> a(n);
    for (int i = 0; i < n; i++) {
        cin >> a[i];
    }

    vector<int> suf_min(n);
    suf_min[n - 1] = a[n - 1];
    for (int i = n - 2; i >= 0; i--) {
        suf_min[i] = min(a[i], suf_min[i + 1]);
    }

    for (int i = 0; i < n; i++) {
        if (i) {
            cout << ' ';
        }
        cout << suf_min[i];
    }
    cout << '\n';

    return 0;
}